service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/CbsSessionHandler.java

```diff
@@ -0,0 +1,193 @@
+import com.azure.core.credential.TokenCredential;
+import com.microsoft.azure.sdk.iot.deps.auth.TokenCredentialType;
+import com.microsoft.azure.sdk.iot.deps.transport.amqp.ErrorLoggingBaseHandlerWithCleanup;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.qpid.proton.amqp.messaging.Accepted;
+import org.apache.qpid.proton.amqp.transport.DeliveryState;
+import org.apache.qpid.proton.engine.BaseHandler;
+import org.apache.qpid.proton.engine.EndpointState;
+import org.apache.qpid.proton.engine.Event;
+import org.apache.qpid.proton.engine.Receiver;
+import org.apache.qpid.proton.engine.Sender;
+import org.apache.qpid.proton.engine.Session;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.UUID;
+
+@Slf4j
+public class CbsSessionHandler extends ErrorLoggingBaseHandlerWithCleanup implements AuthenticationMessageCallback, LinkStateCallback
+{
+    private Session session;
+    private CbsSenderLinkHandler cbsSenderLinkHandler;
+    private CbsReceiverLinkHandler cbsReceiverLinkHandler;
+    private CbsSessionStateCallback cbsSessionStateCallback;
+    private final TokenCredential authenticationTokenProvider;
+    private final TokenCredentialType authorizationType;
+    private boolean senderLinkOpened = false;
+    private boolean receiverLinkOpened = false;
+
+    CbsSessionHandler(Session session, CbsSessionStateCallback cbsSessionStateCallback, TokenCredential authenticationTokenProvider, TokenCredentialType authorizationType)
+    {
+        this.session = session;
+
+        //All events in this reactor that happened to this session will be handled in this instance (onSessionRemoteOpen, for instance)
+        BaseHandler.setHandler(this.session, this);
+
+        this.cbsSessionStateCallback = cbsSessionStateCallback;
+        this.authenticationTokenProvider = authenticationTokenProvider;
+        this.authorizationType = authorizationType;
+        this.session.open();
+    }
+
+    @Override
+    public void onSessionLocalOpen(Event event)
+    {
+        this.session = event.getSession();
+
+        Sender cbsSender = this.session.sender(CbsSenderLinkHandler.getCbsTag());
+        this.cbsSenderLinkHandler = new CbsSenderLinkHandler(cbsSender, this, this.authenticationTokenProvider, this.authorizationType);
+
+        Receiver cbsReceiver = this.session.receiver(CbsReceiverLinkHandler.getCbsTag());
+        this.cbsReceiverLinkHandler = new CbsReceiverLinkHandler(cbsReceiver, this, this);
+    }
+
+    @Override
+    public void onSessionRemoteOpen(Event e)
+    {
+        log.trace("CBS session opened remotely");
+    }
+
+    @Override
+    public void onSessionLocalClose(Event e)
+    {
+        log.trace("CBS session closed remotely");
+        this.session.getConnection().close();
+
+        this.cbsSenderLinkHandler.close();
+        this.cbsReceiverLinkHandler.close();
+    }
+
+    @Override
+    public void onSessionRemoteClose(Event e)
```

Comment:
is there any reconnection logic in the service client?

Reply:
Not yet, no